### PR TITLE
Extend SearchRuns to support datasets

### PR DESCRIPTION
**Pull request by prithvikannan**
## Related Issues/PRs

Based off of https://github.com/mlflow/mlflow/pull/8475

## What changes are proposed in this pull request?

Extend SearchRuns to support datasets in`filter_string`. Also update docs to reflect this new syntax.

Note: had trouble updating https://github.com/mlflow/mlflow/pull/8475 to point to `master` branch, so made this PR instead. This PR incorporates @dbczumar's feedback.

## How is this patch tested?

- [x] Existing unit/integration tests
- [x] New unit/integration tests
- [ ] Manual tests (describe details, including test results, below)

## Does this PR change the documentation?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Make sure the changed pages / sections render correctly in the documentation preview.

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

Extend SearchRuns to support datasets in`filter_string`.

### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [x] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [x] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: docs/source/search-runs.rst [hunk 1]
```diff
@@ -156,6 +156,23 @@ You can search using the following run attributes contained in :py:class:`mlflow
   attributes.run_id = 'a1b2c3d4'
   attributes.run_id IN ('a1b2c3d4', 'e5f6g7h8')
 
+Datasets
+~~~~~~~~~~~~~~
+
+You can search using the following dataset attributes contained in :py:class:`mlflow.entities.Dataset`: ``name``, ``digest``, ``source_type``, ``source``, ``schema``, and ``profile``.
```
Comment (dbczumar) [inline docs/source/search-runs.rst:162]: I think we should start with `name`, `digest`, and `context`. It's unlikely / unclear to me that users would have an entire source / schema / profile dictionary to search against, and I also don't think it's likely that users will need to search based on source_type (can add it later, if necessary).

This PR only seems to test `name`, `digest`, and `context`, so removing these other attributes will save us time :)

File: docs/source/search-runs.rst [hunk 2]
```diff
@@ -156,6 +156,23 @@ You can search using the following run attributes contained in :py:class:`mlflow
...
+
+All dataset attributes are string values. Other fields in :py:class:`mlflow.entities.Dataset` are not searchable.
+
+.. rubric:: Example
+
+.. code-block:: sql
+
+  datasets.name = 'mydataset'
+  datasets.digest = 's8ds293b'
+  datasets.digest IN ('s8ds293b', 'jks834s2')
```
Comment (dbczumar) [inline docs/source/search-runs.rst:174]: Can we include a context tag search in this example?

File: docs/source/search-runs.rst [hunk 3]
```diff
@@ -156,6 +156,23 @@ You can search using the following run attributes contained in :py:class:`mlflow
   attributes.run_id = 'a1b2c3d4'
   attributes.run_id IN ('a1b2c3d4', 'e5f6g7h8')
 
+Datasets
+~~~~~~~~~~~~~~
+
```
Comment (dbczumar) [inline docs/source/search-runs.rst:161]: If it's not too much trouble, can we also add a datasets search example to the UI tooltip that pops up next to the runs search bar on the MLflow Experiment Page UI?

File: mlflow/utils/search_utils.py [hunk 4]
```diff
@@ -113,12 +115,14 @@ class SearchUtils:
...
     VALID_STRING_ATTRIBUTE_COMPARATORS = {"!=", "=", LIKE_OPERATOR, ILIKE_OPERATOR, "IN", "NOT IN"}
     VALID_NUMERIC_ATTRIBUTE_COMPARATORS = VALID_METRIC_COMPARATORS
+    VALID_DATASET_COMPARATORS = {"!=", "=", LIKE_OPERATOR, ILIKE_OPERATOR, "IN", "NOT IN"}
     _BUILTIN_NUMERIC_ATTRIBUTES = {"start_time", "end_time"}
     _ALTERNATE_NUMERIC_ATTRIBUTES = {"created", "Created"}
     _ALTERNATE_STRING_ATTRIBUTES = {"run name", "Run name", "Run Name"}
     NUMERIC_ATTRIBUTES = set(
         list(_BUILTIN_NUMERIC_ATTRIBUTES) + list(_ALTERNATE_NUMERIC_ATTRIBUTES)
     )
+    DATASET_ATTRIBUTES = {"name", "digest", "source_type", "source", "schema", "profile", "context"}
```
Comment (dbczumar) [inline mlflow/utils/search_utils.py:125]: As above (https://github.com/mlflow/mlflow/pull/8622/files#r1220841776), I think we should start with only name, digest, and context as valid attributes

File: mlflow/utils/search_utils.py [hunk 5]
```diff
@@ -113,12 +115,14 @@ class SearchUtils:
     VALID_TAG_COMPARATORS = {"!=", "=", LIKE_OPERATOR, ILIKE_OPERATOR}
     VALID_STRING_ATTRIBUTE_COMPARATORS = {"!=", "=", LIKE_OPERATOR, ILIKE_OPERATOR, "IN", "NOT IN"}
     VALID_NUMERIC_ATTRIBUTE_COMPARATORS = VALID_METRIC_COMPARATORS
+    VALID_DATASET_COMPARATORS = {"!=", "=", LIKE_OPERATOR, ILIKE_OPERATOR, "IN", "NOT IN"}
```
Comment (dbczumar) [inline mlflow/utils/search_utils.py:118]: Do we have test for LIKE / ILIKE? If not, can we add that coverage?

File: mlflow/store/tracking/sqlalchemy_store.py [hunk 6]
```diff
@@ -1497,21 +1511,49 @@ def _get_sqlalchemy_filter_clauses(parsed, session, dialect):
...
+                        .join(
+                            SqlInputTag,
+                            and_(
+                                SqlInputTag.input_uuid == SqlInput.input_uuid,
+                                SqlInputTag.name == MLFLOW_DATASET_CONTEXT,
+                                SqlInputTag.value == value,
+                            ),
+                        )
+                        .subquery()
+                    )
```
Comment (dbczumar) [inline mlflow/store/tracking/sqlalchemy_store.py:1536]: Does this work for comparators other than `=`? Do we have to include `SearchUtils.get_sql_comparison_func` here too?

Can we also add tests for `IN`, `LIKE`, and `ILIKE` on the `context` tag in the file store and sqlalchemy store?

File: docs/source/search-runs.rst (hunk 1 above)
Comment (prithvikannan) [reply docs/source/search-runs.rst]: agreed. the story for searching those other fields doesn't make too much sense for now.

**Review by dbczumar — APPROVED** [review]
LGTM once we add a test for `context` with `IN (...)`. Thanks @prithvikannan !

**Pull request merged**